apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/KmsTest.cs

```diff
@@ -0,0 +1,247 @@
+
+            // We shouldn't be able to download the object without supplying an encryption key.
+            Assert.Throws<GoogleApiException>(() => client.DownloadObject(bucketName, objectName, new MemoryStream()));
+
+            // We should be able to download it again when we supply the encryption key
+            TestHelpers.ValidateData(bucketName, objectName, data, options: new DownloadObjectOptions { EncryptionKey = key });
+        }
+
+        [SkippableFact]
+        public void UploadObject_BucketHasNoDefaultKmsKey_UploadWithKmsKey()
+        {
+            Skip.If(SkipTests);
+            var client = _fixture.Client;
+
+            string bucketName = _fixture.InitiallyEmptyBucket;
+            string objectName = TestHelpers.GenerateName();
+            var data = TestHelpers.GenerateData(1024);
+            EncryptionKey key = EncryptionKey.Generate();
+
+            // Upload an object with a KMS key.
+            Object obj = client.UploadObject(bucketName, objectName, null, data, new UploadObjectOptions { KmsKeyName = KeyName2 });
+            Assert.StartsWith(VersionedKeyName2Prefix, obj.KmsKeyName);
+
+            // We should be able to download it again.
+            TestHelpers.ValidateData(bucketName, objectName, data);
+        }
+
+        [SkippableFact]
+        public void UploadObject_BucketHasDefaultKmsKey_UploadWithClientDefaultCsek()
+        {
+            Skip.If(SkipTests);
+
+            string bucketName = CreateBucketWithDefaultKmsKeyName(KeyName1);
+            string objectName = TestHelpers.GenerateName();
+            var data = TestHelpers.GenerateData(1024);
+            EncryptionKey key = EncryptionKey.Generate();
+            var client = StorageClient.Create(encryptionKey: key);
+
+            // Upload an object using a client that has an encryption key. The default KMS key is ignored.
+            Object obj = client.UploadObject(bucketName, objectName, null, data);
+            Assert.Null(obj.KmsKeyName);
+            Assert.Equal(key.Base64Hash, obj.CustomerEncryption.KeySha256);
+
+            // We shouldn't be able to download the object without supplying an encryption key, e.g. using the fixture's client
+            Assert.Throws<GoogleApiException>(() => _fixture.Client.DownloadObject(bucketName, objectName, new MemoryStream()));
+
+            // We should be able to download it again when we supply the encryption key, either as an option
+            // or via the default encryption key in the client.
+            TestHelpers.ValidateData(bucketName, objectName, data, _fixture.Client, new DownloadObjectOptions { EncryptionKey = key });
+            TestHelpers.ValidateData(bucketName, objectName, data, client);
+        }
+
+        [SkippableFact]
+        public void UploadObject_BucketHasNoDefaultKmsKey_UploadWithClientDefaultCsek_ExplicitKmsKey()
+        {
+            Skip.If(SkipTests);
+
+            string bucketName = CreateBucketWithDefaultKmsKeyName(KeyName1);
```

Comment:
According to the name of the test, this should use `_fixture.InitiallyEmptyBucket` I believe (or the name of the test is wrong).

Reply:
Whoops - yes, you're right. Will fix.